### PR TITLE
[KYUUBI #6249] Drop support for Flink 1.16.

### LINKED ISSUE #6249: [TASK][EASY] Drop support for Flink 1.16
### What's the level of this task?

EASY

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before creating

- [X] I have searched in the [task list](https://github.com/orgs/apache/projects/343) and found no similar tasks.

### Mentor

- [X] I have sufficient expertise on this task, and I volunteer to be a mentor of this task to guide contributors through the task.

### Skill requirements

- Basic knowledge of Scala, Flink, GitHub Actions

### Background and Goals

We have already deprecated the support for Flink 1.16 in 1.9.0(https://github.com/apache/kyuubi/pull/6142), now, we'd like to drop the support for Flink 1.16 and obliterate the related code.

### Implementation steps

- Update code of `kyuubi-flink-sql-engine`
- Update CI(GHA workflow)
- Update docs (quick start, migration guide, etc.)

### Additional context

Introduction of [2024H1 Kyuubi Code Contribution Program](https://github.com/apache/kyuubi/issues/6232)

### PR DESCRIPTION
**Pull request by slfan1989**
# :mag: Description
## Issue References 🔗

This pull request fixes #6249.

## Describe Your Solution 🔧

According to the plan, we will no longer support Flink 1.16, and this PR will try to remove Flink 1.16.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Test Plan 🧪

Pass GHA

---

# Checklist 📝

- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

File: docs/deployment/migration-guide.md [hunk 1]
```diff
@@ -28,7 +28,7 @@
 * Since Kyuubi 1.9.0, the support of Spark SQL engine for Spark 3.1 is deprecated, and will be removed in the future.
 * Since Kyuubi 1.9.0, the support of Spark extensions for Spark 3.1 is removed, please use Spark 3.2 or higher versions.
 * Since Kyuubi 1.9.0, `kyuubi.frontend.login.timeout`, `kyuubi.frontend.thrift.login.timeout`, `kyuubi.frontend.backoff.slot.length`, `kyuubi.frontend.thrift.backoff.slot.length` are removed.
-* Since Kyuubi 1.9.0, the support of Flink engine for Flink 1.16 is deprecated, and will be removed in the future.
+* Since Kyuubi 1.9.0, the support of Flink engine for Flink 1.16 is deprecated and removed.
```
Comment (pan3793) [inline docs/deployment/migration-guide.md:31]: the removal happens in 1.10, please keep this statement and add a new one in 1.9 to 1.10

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/shim/FlinkSessionManager.scala [hunk 2]
```diff
@@ -21,26 +21,17 @@ import org.apache.flink.table.gateway.api.session.{SessionEnvironment, SessionHa
 import org.apache.flink.table.gateway.service.context.DefaultContext
 import org.apache.flink.table.gateway.service.session.Session
 
-import org.apache.kyuubi.engine.flink.FlinkEngineUtils.FLINK_RUNTIME_VERSION
 import org.apache.kyuubi.util.reflect._
 import org.apache.kyuubi.util.reflect.ReflectUtils._
 
 class FlinkSessionManager(engineContext: DefaultContext) {
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/shim/FlinkSessionManager.scala:27]: IIRC this class was introduced as an adapter, is it possible to remove it now?

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/udf/KDFRegistry.scala [hunk 3]
```diff
@@ -37,22 +36,13 @@ object KDFRegistry {
...
     val flinkConfigMap: util.Map[String, String] = {
-      if (FLINK_RUNTIME_VERSION === "1.16") {
-        DynMethods
-          .builder("getConfigMap")
-          .impl(classOf[SessionContext])
-          .build()
-          .invoke(sessionContext)
-          .asInstanceOf[util.Map[String, String]]
-      } else {
         DynMethods
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/udf/KDFRegistry.scala:39]: do we still need reflection now?

**Comment by pan3793** [comment]
@slfan1989 thanks for raising PR quickly, almost good, but it seems there still are some reflection invocations that can be eliminated, and for code style, please run `dev/reformat`

File: docs/deployment/migration-guide.md (hunk 1 above)
Comment (slfan1989) [reply docs/deployment/migration-guide.md:31]: Thank you for reviewing pr! I will improve it.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/shim/FlinkSessionManager.scala (hunk 2 above)
Comment (slfan1989) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/shim/FlinkSessionManager.scala:27]: Thanks for your suggestion! I will try to remove this class.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/udf/KDFRegistry.scala (hunk 3 above)
Comment (slfan1989) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/udf/KDFRegistry.scala:39]: I will improve it.

File: docs/deployment/migration-guide.md [hunk 4]
```diff
@@ -20,6 +20,7 @@
 ## Upgrading from Kyuubi 1.9 to 1.10
 
 * Since Kyuubi 1.10, `beeline` is deprecated and will be removed in the future, please use `kyuubi-beeline` instead.
+* Since Kyuubi 1.10.0, the support of Flink engine for Flink 1.16 is deprecated and removed, please use Flink1.17 and above versions.
```
Comment (pan3793) [inline docs/deployment/migration-guide.md:23]:
```suggestion
* Since Kyuubi 1.10, the support of Flink engine for Flink 1.16 is removed.
```

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merging to master

**Comment by slfan1989** [comment]
@pan3793 Thank you very much for reviewing the code!

**Pull request closed without merging**